File: bindings/swift/Sources/MaplibreNativeFFI/Query.swift
internal import CMaplibreNativeC
import Foundation

public enum RenderedQueryGeometry: Equatable, Sendable {
  case point(ScreenPoint)
  /// Screen-space box in logical map pixels. Corners may be given in any order
  /// and may extend past the viewport; queries normalize and clip them.
  case box(min: ScreenPoint, max: ScreenPoint)
  case lineString([ScreenPoint])

  var nativeGeometry: NativeRenderedQueryGeometry {
    switch self {
    case let .point(point): .point(point.nativeInput)
    case let .box(min, max): .box(min: min.nativeInput, max: max.nativeInput)
    case let .lineString(points): .lineString(points.map(\.nativeInput))
    }
  }
}

public struct RenderedFeatureQueryOptions: Equatable, Sendable {
  public var layerIds: [String]
  public var filter: Data?

  public init(layerIds: [String] = [], filter: Data? = nil) {
    self.layerIds = layerIds
    self.filter = filter
  }

  var nativeOptions: NativeRenderedFeatureQueryOptions {
    NativeRenderedFeatureQueryOptions(layerIds: layerIds, filter: filter)
  }
}

public struct SourceFeatureQueryOptions: Equatable, Sendable {
  public var sourceLayerIds: [String]
  public var filter: Data?

  public init(sourceLayerIds: [String] = [], filter: Data? = nil) {
    self.sourceLayerIds = sourceLayerIds
    self.filter = filter
  }

  var nativeOptions: NativeSourceFeatureQueryOptions {
    NativeSourceFeatureQueryOptions(
      sourceLayerIds: sourceLayerIds,
      filter: filter
    )
  }
}

public struct FeatureStateSelector: Equatable, Sendable {
  public var sourceId: String
  public var sourceLayerId: String?
  public var featureId: String?
  public var stateKey: String?

  public init(
    sourceId: String,
    sourceLayerId: String? = nil,
    featureId: String? = nil,
    stateKey: String? = nil
  ) {
    self.sourceId = sourceId
    self.sourceLayerId = sourceLayerId
    self.featureId = featureId
    self.stateKey = stateKey
  }

  var nativeSelector: NativeFeatureStateSelector {
    NativeFeatureStateSelector(
      sourceId: sourceId,
      sourceLayerId: sourceLayerId,
      featureId: featureId,
      stateKey: stateKey
    )
  }
}

public struct QueriedFeature: Equatable, Sendable {
  public var feature: Data
  public var sourceId: String?
  public var sourceLayerId: String?
  public var state: Data?

  public init(
    feature: Data,
    sourceId: String? = nil,
    sourceLayerId: String? = nil,
    state: Data? = nil
  ) {
    self.feature = feature
    self.sourceId = sourceId
    self.sourceLayerId = sourceLayerId
    self.state = state
  }
}

public extension RenderSessionHandle {
  func queryRenderedFeatures(
    geometry: RenderedQueryGeometry,
    options: RenderedFeatureQueryOptions = RenderedFeatureQueryOptions()
  ) throws -> [QueriedFeature] {
    try mapNativeFailure {
      try geometry.nativeGeometry.withNativeGeometry { nativeGeometry in
        try options.nativeOptions.withNativeOptions { nativeOptions in
          try NativeQuery.renderedFeatures(
            session: requireLiveHandle(),
            geometry: nativeGeometry,
            options: nativeOptions
          )
        }
      }
    }
  }

  func querySourceFeatures(
    sourceId: String,
    options: SourceFeatureQueryOptions = SourceFeatureQueryOptions()
  ) throws -> [QueriedFeature] {
    try mapNativeFailure {
      let arena = NativeInputArena()
      defer { withExtendedLifetime(arena) {} }
      return try options.nativeOptions.withNativeOptions { nativeOptions in
        try NativeQuery.sourceFeatures(
          session: self.requireLiveHandle(),
          sourceId: arena.view(sourceId),
          options: nativeOptions
        )
      }
    }
  }

  func queryFeatureExtension(
    sourceId: String,
    feature: Data,
    extensionName: String,
    extensionField: String,
    arguments: Data? = nil
  ) throws -> Data {
    try mapNativeFailure {
      let arena = NativeInputArena()
      defer { withExtendedLifetime(arena) {} }
      let call = { (arguments: UnsafePointer<mln_buffer_view>?) throws in
        try NativeQuery.featureExtensions(
          session: self.requireLiveHandle(),
          sourceId: arena.view(sourceId),
          feature: arena.view(feature),
          extensionName: arena.view(extensionName),
          extensionField: arena.view(extensionField),
          arguments: arguments
        )
      }
      guard let arguments else { return try call(nil) }
      var argumentsView = arena.view(arguments)
      return try withUnsafePointer(to: &argumentsView, call)
    }
  }
}

File: bindings/python/python/maplibre_native_ffi/query.py
"""Rendered and source query descriptors."""

from __future__ import annotations

from dataclasses import dataclass

from ._enum import NativeIntEnum
from .camera import ScreenPoint


class RenderedQueryGeometryType(NativeIntEnum):
    """Rendered feature query geometry variants."""

    POINT = 1
    BOX = 2
    LINE_STRING = 3


@dataclass(frozen=True, slots=True)
class ScreenBox:
    """Screen-space box in logical map pixels.

    Corners may be given in any order and may extend past the viewport;
    rendered queries normalize and clip them.
    """

    min: ScreenPoint
    max: ScreenPoint


@dataclass(frozen=True, slots=True)
class RenderedQueryGeometry:
    """Screen-space geometry used for rendered feature queries."""

    type: RenderedQueryGeometryType
    point: ScreenPoint | None = None
    box: ScreenBox | None = None
    line_string: tuple[ScreenPoint, ...] | None = None

    @classmethod
    def point_geometry(cls, point: ScreenPoint) -> RenderedQueryGeometry:
        """Create a point query geometry."""
        return cls(RenderedQueryGeometryType.POINT, point=point)

    @classmethod
    def box_geometry(cls, box_: ScreenBox) -> RenderedQueryGeometry:
        """Create a box query geometry."""
        return cls(RenderedQueryGeometryType.BOX, box=box_)

    @classmethod
    def line_string_geometry(
        cls, points: list[ScreenPoint] | tuple[ScreenPoint, ...]
    ) -> RenderedQueryGeometry:
        """Create a line-string query geometry."""
        return cls(RenderedQueryGeometryType.LINE_STRING, line_string=tuple(points))

    def __post_init__(self) -> None:
        active = sum(
            value is not None for value in (self.point, self.box, self.line_string)
        )
        if active != 1:
            raise ValueError(
                "rendered query geometry must contain exactly one geometry value"
            )
        if self.type is RenderedQueryGeometryType.POINT and self.point is None:
            raise ValueError("point query geometry requires point")
        if self.type is RenderedQueryGeometryType.BOX and self.box is None:
            raise ValueError("box query geometry requires box")
        if (
            self.type is RenderedQueryGeometryType.LINE_STRING
            and self.line_string is None
        ):
            raise ValueError("line-string query geometry requires points")


@dataclass(frozen=True, slots=True)
class RenderedFeatureQueryOptions:
    """Options for rendered feature queries."""

    layer_ids: tuple[str, ...] | None = None
    filter: bytes | None = None


@dataclass(frozen=True, slots=True)
class SourceFeatureQueryOptions:
    """Options for source feature queries."""

    source_layer_ids: tuple[str, ...] | None = None
    filter: bytes | None = None


@dataclass(frozen=True, slots=True)
class FeatureStateSelector:
    """Source, feature, and state-key selector for render-session feature state."""

    source_id: str
    source_layer_id: str | None = None
    feature_id: str | None = None
    state_key: str | None = None

    def __post_init__(self) -> None:
        if self.state_key is not None and self.feature_id is None:
            raise ValueError("state_key requires feature_id")


def _point_to_native_wire(point: ScreenPoint) -> tuple[float, float]:
    return (point.x, point.y)


def _geometry_to_native_wire(geometry: RenderedQueryGeometry) -> dict[str, object]:
    if geometry.type is RenderedQueryGeometryType.POINT:
        if geometry.point is None:
            raise ValueError("point query geometry requires point")
        return {"type": "point", "point": _point_to_native_wire(geometry.point)}
    if geometry.type is RenderedQueryGeometryType.BOX:
        if geometry.box is None:
            raise ValueError("box query geometry requires box")
        return {
            "type": "box",
            "min": _point_to_native_wire(geometry.box.min),
            "max": _point_to_native_wire(geometry.box.max),
        }
    if geometry.type is RenderedQueryGeometryType.LINE_STRING:
        if geometry.line_string is None:
            raise ValueError("line-string query geometry requires points")
        return {
            "type": "line_string",
            "points": [_point_to_native_wire(point) for point in geometry.line_string],
        }
    raise TypeError(f"unknown rendered query geometry type: {geometry.type!r}")


__all__ = [
    "FeatureStateSelector",
    "RenderedFeatureQueryOptions",
    "RenderedQueryGeometry",
    "RenderedQueryGeometryType",
    "ScreenBox",
    "SourceFeatureQueryOptions",
]
